File: bin/method_annotation-cellassign.py
#!/usr/bin/env python3

"""
Feature selection

Usage:
    method-annotation-cellassign.py [options] <file>
    
Options:
    -h --help               Show this screen.
"""

def run_cellassign(adata, mrk):
    """
    Run the scanpy core analysis

    Parameters
    ----------
    adata : AnnData

    Returns
    -------
    CellAssign model.

    """
    import scvi
    
    # Filter data
    bdata = adata[:, mrk.index].copy()
    # del adata
    
    # Use raw counts
    bdata.X = bdata.layers["counts"]
    
    # Set up object
    scvi.external.CellAssign.setup_anndata(bdata, 
                                           size_factor_key="size_factor")
    
    # Train model
    model = scvi.external.CellAssign(bdata, mrk)
    model.train()
    
    return(model)


def main():
    """The main script function"""
    from docopt import docopt
    import scanpy as sc
    import pandas as pd
    import numpy as np

    # Variables
    args = docopt(__doc__)
    
    in_file = args["<file>"]
    in_markers = args["<markers>"]
    out_file = in_file
    
    print(f"Reading dataset from '{in_file}'")
    ds = sc.read(in_file)

    print(f"Reading markers from '{in_markers}'")
    markers = pd.read_excel(in_markers, "level_3")
    mrk = pd.DataFrame()
    mrk["type"] = markers["type"]
    mrk.index = markers["gene"].values
    mrk = pd.get_dummies(mrk)
    
    if "size_factor" in ds.obs:
        print("Size factor already present.")
    else:
        print("Adding size factor...")
        lib_size = ds.obs["libsize"].values
        ds.obs["size_factor"]= lib_size / np.mean(lib_size)
    
    print("Running CellAssign...")
    label = run_cellassign(ds, mrk)

    # Write data
    print(f"Writing output to '{out_file}'")
    ds.write_h5ad(out_file)

    print("Done.")

if __name__ == "__main__":
    main()

File: bin/method_annotation-cellassign.R
"
Cell type annotation

Usage:
    analysis_annotation-cellassign.R [options] <file>
    
Options:
    -h --help             Show this screen.
" -> doc

# Source functions
suppressMessages({
  for (i in list.files("bin/functions", full.names = TRUE)) {
    source(i)
  }
})

#' The main script function
main <- function() {
  
  # Variables ------------------------------------------------------------------
  args <- docopt::docopt(doc)
  
  in_file <- args[["<file>"]]
  
  out_file_ann <- stringr::str_replace(out_file, ".h5ad", "_cellassign.csv")
  out_dir <- stringr::str_replace(
    stringr::str_replace(in_file, "data", "analysis"), ".h5ad", "/"
  )
  if (!dir.exists(out_dir)) {
    dir.create(out_dir, recursive = TRUE)
  }
  
  # Read data ------------------------------------------------------------------
  message("Reading data ...")
  ds <- read_h5ad(in_file, layers = "counts")
  
  # Marker genes
  xls_url <- "https://nubes.helmholtz-berlin.de/s/id9rrKNeditKMmF"
  xls_url <- paste0(xls_url, "/", "download")
  xls <- "docs/celltype/markers.xlsx"
  download.file(xls_url, xls)
  markers <- list()
  for (i in readxl::excel_sheets(xls)) {
    markers[[i]] <- readxl::read_excel(xls, sheet = i)
  }
  
  # Run CellAssign -------------------------------------------------------------
  message("Running CellAssign...")
  
  mrk <- markers$level_3
  mrk <- split(mrk$gene, mrk$type)
  mrk <- cellassign::marker_list_to_mat(mrk)
  
  genes <- rownames(mrk)
  genes <- genes[sparseMatrixStats::rowSums2(ds[genes, ]@assays@data$X)>0]
  
  cells <- colnames(ds)
  cells <- cells[sparseMatrixStats::colSums2(ds[genes, ]@assays@data$X)>0]
  
  mrk <- mrk[genes, ]
  label <- cellassign::cellassign(ds[genes, cells], mrk, sce_assay = "X")
  
  cellassign::cellassign(
    cellassign::example_sce, cellassign::example_rho
  )
  
  message("Done!")
}

if (sys.nframe() == 0) {
  main()
}
